File: baekjoon/graph/2252/line.py
import sys
from collections import deque

def topological_sort(v_cnt, linked, indegree):
    ans = []
    q = deque()

    for v in range(1, v_cnt):
        if indegree[v] == 0:
            q.append(v)

    while q:
        cur = q.popleft()
        ans.append(cur)
        for next in linked[cur]:
            indegree[next] -= 1
            if indegree[next] == 0:
                q.append(next)

    for cur in ans:
        print(cur, end=' ')
    print()

if __name__ == '__main__':
    v_cnt, e_cnt = map(int, sys.stdin.readline().split(' '))
    v_cnt += 1

    linked = [[] for _ in range(v_cnt + 1)]
    indegree = [0 for _ in range(v_cnt + 1)]
    for _ in range(e_cnt):
        _from, to = map(int, sys.stdin.readline().split(' '))
        linked[_from].append(to)
        indegree[to] += 1

    topological_sort(v_cnt, linked, indegree)

File: baekjoon/graph/2252/line.cpp
#include <iostream>
#include <vector>

typedef std::vector<int> t_vec;

const int   MAX = 32001;

t_vec map[MAX];
bool  visited[MAX];
t_vec ans;
int   v_cnt;

void    dfs(int cur_v)
{
    visited[cur_v] = true;

    t_vec   &linked = map[cur_v];
    for (const int &next : linked)
    {
        if (!visited[next])
            dfs(next);
    }

    ans.push_back(cur_v);
}

void    topological_sort(void)
{
    for (int v = 1; v <= v_cnt; ++v)
    {
        if (!visited[v])
            dfs(v);
    }

    for (int i = ans.size() - 1; i >= 0; --i)
        std::cout << ans[i] << ' ';
    std::cout << '\n';
}

int main(void)
{
    std::cin.tie(0)->sync_with_stdio(0);

    int e_cnt;

    std::cin >> v_cnt >> e_cnt;
    while (e_cnt--)
    {
        int from, to;
        std::cin >> from >> to;
        map[from].push_back(to);
    }

    topological_sort();
    return (0);
}
